// @renderinc/sdk/dist/workflows/client/client.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.WorkflowsClient = void 0;
const eventsource_1 = require("eventsource");
const errors_js_1 = require("../../errors.js");
const version_js_1 = require("../../version.js");
const sse_js_1 = require("./sse.js");
const task_run_result_js_1 = require("./task-run-result.js");
function handleApiError(error, response, context) {
    const statusCode = response.status;
    const errorMessage = `${context}: ${error}`;
    if (statusCode >= 500) {
        throw new errors_js_1.ServerError(errorMessage, statusCode, error);
    }
    else if (statusCode >= 400) {
        throw new errors_js_1.ClientError(errorMessage, statusCode, error);
    }
    throw new errors_js_1.ClientError(errorMessage, statusCode, error);
}
class WorkflowsClient {
    constructor(apiClient, baseUrl, token) {
        this.apiClient = apiClient;
        this.baseUrl = baseUrl;
        this.token = token;
    }
    async *taskRunEvents(taskRunIds, signal) {
        if (signal?.aborted) {
            throw new errors_js_1.AbortError();
        }
        const queue = [];
        let resolve = null;
        let finished = false;
        let streamError = null;
        const push = (item) => {
            queue.push(item);
            if (resolve) {
                resolve();
                resolve = null;
            }
        };
        const fail = (err) => {
            streamError = err;
            finished = true;
            if (resolve) {
                resolve();
                resolve = null;
            }
        };
        const url = new URL("/v1/task-runs/events", this.baseUrl);
        url.searchParams.append("taskRunIds", taskRunIds.join(","));
        const eventSource = new eventsource_1.EventSource(url.toString(), {
            fetch: (input, init) => fetch(input, {
                ...init,
                headers: {
                    ...init?.headers,
                    Authorization: `Bearer ${this.token}`,
                    "User-Agent": (0, version_js_1.getUserAgent)(),
                },
            }),
        });
        const eventHandler = (event) => {
            try {
                const details = JSON.parse(event.data);
                push(details);
            }
            catch (e) {
                fail(new Error(`Failed to parse task run details: ${e}`));
            }
        };
        const errorHandler = (error) => {
            fail(new Error(`SSE connection error: ${error.message || "Unknown error"}`));
        };
        const abortHandler = () => {
            cleanup();
            fail(new errors_js_1.AbortError());
        };
        const cleanup = () => {
            eventSource.removeEventListener(sse_js_1.TaskEventType.COMPLETED, eventHandler);
            eventSource.removeEventListener(sse_js_1.TaskEventType.FAILED, eventHandler);
            eventSource.removeEventListener(sse_js_1.TaskEventType.CANCELED, eventHandler);
            eventSource.removeEventListener("error", errorHandler);
            eventSource.close();
            signal?.removeEventListener("abort", abortHandler);
        };
        eventSource.addEventListener(sse_js_1.TaskEventType.COMPLETED, eventHandler);
        eventSource.addEventListener(sse_js_1.TaskEventType.FAILED, eventHandler);
        eventSource.addEventListener(sse_js_1.TaskEventType.CANCELED, eventHandler);
        eventSource.addEventListener("error", errorHandler);
        signal?.addEventListener("abort", abortHandler);
        try {
            while (true) {
                while (queue.length > 0) {
                    yield queue.shift();
                }
                if (finished) {
                    break;
                }
                await new Promise((r) => {
                    resolve = r;
                });
            }
            if (streamError) {
                throw streamError;
            }
        }
        finally {
            cleanup();
        }
    }
    async startTask(taskSlug, inputData, signal) {
        if (signal?.aborted) {
            throw new errors_js_1.AbortError();
        }
        try {
            const { data, error, response } = await this.apiClient.POST("/task-runs", {
                body: {
                    task: taskSlug,
                    input: inputData,
                },
                signal,
            });
            if (error) {
                handleApiError(error, response, "Failed to run task");
            }
            return new task_run_result_js_1.TaskRunResult((id, sig) => this.waitOnTaskRun(id, sig), data.id, signal);
        }
        catch (err) {
            if (err instanceof DOMException && err.name === "AbortError") {
                throw new errors_js_1.AbortError();
            }
            throw err;
        }
    }
    async runTask(taskSlug, inputData, signal) {
        const result = await this.startTask(taskSlug, inputData, signal);
        return result.get();
    }
    async getTaskRun(taskRunId) {
        const { data, error, response } = await this.apiClient.GET("/task-runs/{taskRunId}", {
            params: { path: { taskRunId } },
        });
        if (error) {
            handleApiError(error, response, "Failed to get task run");
        }
        return data;
    }
    async cancelTaskRun(taskRunId) {
        const { error, response } = await this.apiClient.DELETE("/task-runs/{taskRunId}", {
            params: { path: { taskRunId } },
        });
        if (error) {
            handleApiError(error, response, "Failed to cancel task run");
        }
    }
    async listTaskRuns(params) {
        const { data, error, response } = await this.apiClient.GET("/task-runs", {
            params: { query: params },
        });
        if (error) {
            handleApiError(error, response, "Failed to list task runs");
        }
        return data;
    }
    async waitOnTaskRun(taskRunId, signal) {
        for await (const event of this.taskRunEvents([taskRunId], signal)) {
            return event;
        }
        throw new Error(`SSE stream ended without receiving an event for task run ${taskRunId}`);
    }
}
exports.WorkflowsClient = WorkflowsClient;
