// @renderinc/sdk/dist/experimental/object/api.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ObjectApi = void 0;
const errors_js_1 = require("./errors.js");
class ObjectApi {
    constructor(apiClient) {
        this.apiClient = apiClient;
    }
    async getUploadUrl(ownerId, region, key, sizeBytes) {
        const { data, error, response } = await this.apiClient.PUT("/objects/{ownerId}/{region}/{key}", {
            params: { path: { ownerId, region: region, key } },
            body: { sizeBytes },
        });
        if (error) {
            (0, errors_js_1.throwObjectApiError)("Failed to get upload URL", response, error);
        }
        return {
            url: data.url,
            expiresAt: new Date(data.expiresAt),
            maxSizeBytes: data.maxSizeBytes,
        };
    }
    async getDownloadUrl(ownerId, region, key) {
        const { data, error, response } = await this.apiClient.GET("/objects/{ownerId}/{region}/{key}", {
            params: { path: { ownerId, region: region, key } },
        });
        if (error) {
            (0, errors_js_1.throwObjectApiError)("Failed to get download URL", response, error);
        }
        return {
            url: data.url,
            expiresAt: new Date(data.expiresAt),
        };
    }
    async delete(ownerId, region, key) {
        const { error, response } = await this.apiClient.DELETE("/objects/{ownerId}/{region}/{key}", {
            params: { path: { ownerId, region: region, key } },
        });
        if (error) {
            (0, errors_js_1.throwObjectApiError)("Failed to delete object", response, error);
        }
    }
    async listObjects(ownerId, region, cursor, limit) {
        const { data, error, response } = await this.apiClient.GET("/objects/{ownerId}/{region}", {
            params: {
                path: { ownerId, region: region },
                query: { cursor, limit },
            },
        });
        if (error) {
            (0, errors_js_1.throwObjectApiError)("Failed to list objects", response, error);
        }
        const objects = data.items.map((item) => ({
            key: item.object.key,
            size: item.object.sizeBytes,
            lastModified: new Date(item.object.lastModified),
        }));
        return { objects, hasNext: data.hasNext, nextCursor: data.nextCursor };
    }
}
exports.ObjectApi = ObjectApi;
