// @renderinc/sdk/dist/experimental/object/client.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ScopedObjectClient = exports.ObjectClient = void 0;
const node_stream_1 = require("node:stream");
const errors_js_1 = require("../../errors.js");
const errors_js_2 = require("./errors.js");
class ObjectClient {
    constructor(apiClient, defaultOwnerId, defaultRegion) {
        this.apiClient = apiClient;
        this.defaultOwnerId = defaultOwnerId;
        this.defaultRegion = defaultRegion;
    }
    resolveOwnerId(ownerId) {
        const resolved = ownerId || this.defaultOwnerId;
        if (!resolved) {
            throw new errors_js_1.RenderError("ownerId is required. Provide it as a parameter or set the RENDER_WORKSPACE_ID environment variable.");
        }
        return resolved;
    }
    resolveRegion(region) {
        const resolved = region || this.defaultRegion;
        if (!resolved) {
            throw new errors_js_1.RenderError("region is required. Provide it as a parameter or set the RENDER_REGION environment variable.");
        }
        return resolved;
    }
    async put(input) {
        const ownerId = this.resolveOwnerId(input.ownerId);
        const region = this.resolveRegion(input.region);
        const size = this.resolveSize(input);
        const { data, error, response: apiResponse, } = await this.apiClient.PUT("/objects/{ownerId}/{region}/{key}", {
            params: {
                path: {
                    ownerId,
                    region: region,
                    key: input.key,
                },
            },
            body: { sizeBytes: size },
        });
        if (error) {
            (0, errors_js_2.throwObjectApiError)("Failed to get upload URL", apiResponse, error);
        }
        if (size !== data.maxSizeBytes) {
            throw new errors_js_1.ClientError(`File size ${size} bytes does not match expected size of ${data.maxSizeBytes} bytes`, 400);
        }
        const headers = {
            "Content-Length": size.toString(),
        };
        if (input.contentType) {
            headers["Content-Type"] = input.contentType;
        }
        let body = input.data;
        if (process.versions.bun && input.data instanceof node_stream_1.Readable) {
            const chunks = [];
            for await (const chunk of input.data) {
                chunks.push(Buffer.from(chunk));
            }
            body = Buffer.concat(chunks);
        }
        const response = await fetch(data.url, {
            method: "PUT",
            headers,
            body,
            duplex: "half",
        });
        if (!response.ok) {
            (0, errors_js_2.throwStorageError)("Upload failed", response);
        }
        return {
            etag: response.headers.get("ETag") ?? undefined,
        };
    }
    async get(input) {
        const ownerId = this.resolveOwnerId(input.ownerId);
        const region = this.resolveRegion(input.region);
        const { data, error, response: apiResponse, } = await this.apiClient.GET("/objects/{ownerId}/{region}/{key}", {
            params: {
                path: {
                    ownerId,
                    region: region,
                    key: input.key,
                },
            },
        });
        if (error) {
            (0, errors_js_2.throwObjectApiError)("Failed to get download URL", apiResponse, error);
        }
        const response = await fetch(data.url);
        if (!response.ok) {
            (0, errors_js_2.throwStorageError)("Download failed", response);
        }
        const arrayBuffer = await response.arrayBuffer();
        const buffer = Buffer.from(arrayBuffer);
        return {
            data: buffer,
            size: buffer.byteLength,
            contentType: response.headers.get("Content-Type") ?? undefined,
        };
    }
    async delete(input) {
        const ownerId = this.resolveOwnerId(input.ownerId);
        const region = this.resolveRegion(input.region);
        const { error, response } = await this.apiClient.DELETE("/objects/{ownerId}/{region}/{key}", {
            params: {
                path: {
                    ownerId,
                    region: region,
                    key: input.key,
                },
            },
        });
        if (error) {
            (0, errors_js_2.throwObjectApiError)("Failed to delete object", response, error);
        }
    }
    async list(input) {
        const ownerId = this.resolveOwnerId(input.ownerId);
        const region = this.resolveRegion(input.region);
        const { data, error, response } = await this.apiClient.GET("/objects/{ownerId}/{region}", {
            params: {
                path: {
                    ownerId,
                    region: region,
                },
                query: {
                    cursor: input.cursor,
                    limit: input.limit,
                },
            },
        });
        if (error) {
            (0, errors_js_2.throwObjectApiError)("Failed to list objects", response, error);
        }
        const objects = data.items.map((item) => ({
            key: item.object.key,
            size: item.object.sizeBytes,
            lastModified: new Date(item.object.lastModified),
        }));
        return { objects, hasNext: data.hasNext, nextCursor: data.nextCursor };
    }
    scoped(scope) {
        return new ScopedObjectClient(this.apiClient, scope);
    }
    resolveSize(input) {
        if (Buffer.isBuffer(input.data) || input.data instanceof Uint8Array) {
            const actualSize = input.data.byteLength;
            if (input.size !== undefined && input.size !== actualSize) {
                throw new errors_js_1.RenderError(`Size mismatch: provided size ${input.size} does not match actual size ${actualSize}`);
            }
            return actualSize;
        }
        if (typeof input.data === "string") {
            const actualSize = Buffer.byteLength(input.data);
            if (input.size !== undefined && input.size !== actualSize) {
                throw new errors_js_1.RenderError(`Size mismatch: provided size ${input.size} does not match actual size ${actualSize}`);
            }
            return actualSize;
        }
        if (input.size === undefined) {
            throw new errors_js_1.RenderError("Size is required for stream inputs. Provide the size parameter.");
        }
        if (input.size < 0) {
            throw new errors_js_1.RenderError("Size must be a non-negative integer");
        }
        return input.size;
    }
}
exports.ObjectClient = ObjectClient;
class ScopedObjectClient {
    constructor(apiClient, scope) {
        this.scope = scope;
        this.objectClient = new ObjectClient(apiClient);
    }
    async put(input) {
        return this.objectClient.put({
            ...this.scope,
            ...input,
        });
    }
    async get(input) {
        return this.objectClient.get({
            ...this.scope,
            ...input,
        });
    }
    async delete(input) {
        return this.objectClient.delete({
            ...this.scope,
            ...input,
        });
    }
    async list(input = {}) {
        return this.objectClient.list({
            ...this.scope,
            ...input,
        });
    }
}
exports.ScopedObjectClient = ScopedObjectClient;
