// @renderinc/sdk/dist/workflows/uds.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.UDSClient = void 0;
const node_http_1 = __importDefault(require("node:http"));
const version_js_1 = require("../version.js");
const UDS_MAX_RETRIES = 15;
const UDS_INITIAL_DELAY_MS = 250;
const UDS_BACKOFF_FACTOR = 2;
const UDS_MAX_DELAY_MS = 16000;
const CLIENT_ERROR_RE = /^HTTP 4\d{2}:/;
const RETRY_ERROR_RE = /^HTTP 429:/;
class UDSClient {
    constructor(socketPath) {
        this.socketPath = socketPath;
    }
    async getInput() {
        return this.request("/input", "GET");
    }
    buildCallbackBody(results, error) {
        if (results !== undefined) {
            const resultsArray = [results];
            const output = Buffer.from(JSON.stringify(resultsArray)).toString("base64");
            return {
                complete: {
                    output,
                },
            };
        }
        if (error === undefined) {
            throw new Error("Either results or error must be provided");
        }
        return {
            error: {
                details: error,
            },
        };
    }
    async sendCallback(results, error) {
        await this.request("/callback", "POST", this.buildCallbackBody(results, error));
    }
    async runSubtask(taskName, input) {
        const inputBase64 = Buffer.from(JSON.stringify(input)).toString("base64");
        const body = {
            task_name: taskName,
            input: inputBase64,
        };
        const response = await this.request("/run-subtask", "POST", body);
        return response.task_run_id;
    }
    async getSubtaskResult(subtaskId) {
        const body = {
            task_run_id: subtaskId,
        };
        return this.request("/get-subtask-result", "POST", body);
    }
    async registerTasks(tasks) {
        const body = {
            tasks: tasks.map((task) => ({
                name: task.name,
                options: task.options,
            })),
        };
        await this.request("/register-tasks", "POST", body);
    }
    async request(path, method, body) {
        const bodyString = body ? JSON.stringify(body) : "";
        let lastError = new Error("UDS request failed");
        for (let attempt = 0; attempt < UDS_MAX_RETRIES; attempt++) {
            try {
                return await this.requestOnce(path, method, bodyString);
            }
            catch (e) {
                lastError = e instanceof Error ? e : new Error(String(e));
                if (CLIENT_ERROR_RE.test(lastError.message) && !RETRY_ERROR_RE.test(lastError.message)) {
                    throw lastError;
                }
                if (attempt < UDS_MAX_RETRIES - 1) {
                    const delay = Math.min(UDS_INITIAL_DELAY_MS * UDS_BACKOFF_FACTOR ** attempt, UDS_MAX_DELAY_MS);
                    console.warn(`Request to Render failed (attempt ${attempt + 1}/${UDS_MAX_RETRIES}), ` +
                        `retrying in ${delay}ms: ${lastError.message}`);
                    await new Promise((resolve) => setTimeout(resolve, delay));
                }
            }
        }
        throw lastError;
    }
    async requestOnce(path, method, bodyString) {
        return new Promise((resolve, reject) => {
            const req = node_http_1.default.request({
                socketPath: this.socketPath,
                path: path,
                method: method,
                headers: {
                    "Content-Length": Buffer.byteLength(bodyString),
                    "Content-Type": "application/json",
                    "User-Agent": (0, version_js_1.getUserAgent)(),
                },
            }, async (res) => {
                const chunks = [];
                for await (const chunk of res)
                    chunks.push(chunk);
                const responseBody = Buffer.concat(chunks).toString();
                if (res.statusCode && res.statusCode >= 400) {
                    reject(new Error(`HTTP ${res.statusCode}: ${responseBody}`));
                    return;
                }
                if (responseBody.length === 0) {
                    resolve(undefined);
                    return;
                }
                try {
                    resolve(JSON.parse(responseBody));
                }
                catch (error) {
                    reject(error);
                }
            });
            req.on("error", (error) => {
                reject(error);
            });
            req.end(bodyString);
        });
    }
}
exports.UDSClient = UDSClient;
