// @renderinc/sdk/dist/workflows/workflows.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Workflows = void 0;
const errors_js_1 = require("../errors.js");
const registry_js_1 = require("./registry.js");
const runner_js_1 = require("./runner.js");
const task_js_1 = require("./task.js");
class Workflows {
    constructor(options) {
        this._registry = new registry_js_1.TaskRegistry();
        this._defaultRetry = options?.defaultRetry;
        this._defaultTimeout = options?.defaultTimeout;
        this._defaultPlan = options?.defaultPlan;
    }
    get registry() {
        return this._registry;
    }
    task(options, func) {
        const mergedOptions = {
            ...options,
            retry: options.retry ?? this._defaultRetry,
            timeoutSeconds: options.timeoutSeconds ?? this._defaultTimeout,
            plan: options.plan ?? this._defaultPlan,
        };
        this._registry.register(func, mergedOptions);
        return ((...args) => {
            const context = (0, task_js_1.getCurrentContext)();
            if (!context) {
                return func(...args);
            }
            const result = context.executeTask(func, options.name, ...args);
            return result.get();
        });
    }
    async start() {
        await (0, runner_js_1.startTaskServer)(this._registry);
    }
    static fromWorkflows(apps, options) {
        const combined = new Workflows(options);
        for (const app of apps) {
            for (const task of app._registry.getAllTasks()) {
                if (combined._registry.has(task.name)) {
                    throw new errors_js_1.RenderError(`Duplicate task name '${task.name}' when combining workflows`);
                }
                combined._registry.registerMetadata(task);
            }
        }
        return combined;
    }
}
exports.Workflows = Workflows;
