// @renderinc/sdk/dist/experimental/blob/api.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.BlobApi = void 0;
const errors_js_1 = require("../../errors.js");
class BlobApi {
    constructor(apiClient) {
        this.apiClient = apiClient;
    }
    async getUploadUrl(ownerId, region, key, sizeBytes) {
        const { data, error } = await this.apiClient.PUT("/blobs/{ownerId}/{region}/{key}", {
            params: { path: { ownerId, region: region, key } },
            body: { sizeBytes },
        });
        if (error) {
            throw new errors_js_1.RenderError(`Failed to get upload URL: ${error.message || "Unknown error"}`);
        }
        return {
            url: data.url,
            expiresAt: new Date(data.expiresAt),
            maxSizeBytes: data.maxSizeBytes,
        };
    }
    async getDownloadUrl(ownerId, region, key) {
        const { data, error } = await this.apiClient.GET("/blobs/{ownerId}/{region}/{key}", {
            params: { path: { ownerId, region: region, key } },
        });
        if (error) {
            throw new errors_js_1.RenderError(`Failed to get download URL: ${error.message || "Unknown error"}`);
        }
        return {
            url: data.url,
            expiresAt: new Date(data.expiresAt),
        };
    }
    async delete(ownerId, region, key) {
        const { error } = await this.apiClient.DELETE("/blobs/{ownerId}/{region}/{key}", {
            params: { path: { ownerId, region: region, key } },
        });
        if (error) {
            throw new errors_js_1.RenderError(`Failed to delete blob: ${error.message || "Unknown error"}`);
        }
    }
}
exports.BlobApi = BlobApi;
