// @renderinc/sdk/dist/experimental/blob/client.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ScopedBlobClient = exports.BlobClient = void 0;
const errors_js_1 = require("../../errors.js");
class BlobClient {
    constructor(apiClient) {
        this.apiClient = apiClient;
    }
    async put(input) {
        const size = this.resolveSize(input);
        const { data, error } = await this.apiClient.PUT("/blobs/{ownerId}/{region}/{key}", {
            params: {
                path: {
                    ownerId: input.ownerId,
                    region: input.region,
                    key: input.key,
                },
            },
            body: { sizeBytes: size },
        });
        if (error) {
            throw new errors_js_1.RenderError(`Failed to get upload URL: ${error.message || "Unknown error"}`);
        }
        const headers = {
            "Content-Length": size.toString(),
        };
        if (input.contentType) {
            headers["Content-Type"] = input.contentType;
        }
        const response = await fetch(data.url, {
            method: "PUT",
            headers,
            body: input.data,
            duplex: "half",
        });
        if (!response.ok) {
            throw new errors_js_1.RenderError(`Upload failed: ${response.status} ${response.statusText}`);
        }
        return {
            etag: response.headers.get("ETag") ?? undefined,
        };
    }
    async get(input) {
        const { data, error } = await this.apiClient.GET("/blobs/{ownerId}/{region}/{key}", {
            params: {
                path: {
                    ownerId: input.ownerId,
                    region: input.region,
                    key: input.key,
                },
            },
        });
        if (error) {
            throw new errors_js_1.RenderError(`Failed to get download URL: ${error.message || "Unknown error"}`);
        }
        const response = await fetch(data.url);
        if (!response.ok) {
            throw new errors_js_1.RenderError(`Download failed: ${response.status} ${response.statusText}`);
        }
        const arrayBuffer = await response.arrayBuffer();
        const buffer = Buffer.from(arrayBuffer);
        return {
            data: buffer,
            size: buffer.byteLength,
            contentType: response.headers.get("Content-Type") ?? undefined,
        };
    }
    async delete(input) {
        const { error } = await this.apiClient.DELETE("/blobs/{ownerId}/{region}/{key}", {
            params: {
                path: {
                    ownerId: input.ownerId,
                    region: input.region,
                    key: input.key,
                },
            },
        });
        if (error) {
            throw new errors_js_1.RenderError(`Failed to delete blob: ${error.message || "Unknown error"}`);
        }
    }
    scoped(scope) {
        return new ScopedBlobClient(this.apiClient, scope);
    }
    resolveSize(input) {
        if (Buffer.isBuffer(input.data) || input.data instanceof Uint8Array) {
            const actualSize = input.data.byteLength;
            if (input.size !== undefined && input.size !== actualSize) {
                throw new errors_js_1.RenderError(`Size mismatch: provided size ${input.size} does not match actual size ${actualSize}`);
            }
            return actualSize;
        }
        if (input.size === undefined) {
            throw new errors_js_1.RenderError("Size is required for stream and string inputs. Provide the size parameter.");
        }
        if (input.size <= 0) {
            throw new errors_js_1.RenderError("Size must be a positive integer");
        }
        return input.size;
    }
}
exports.BlobClient = BlobClient;
class ScopedBlobClient {
    constructor(apiClient, scope) {
        this.scope = scope;
        this.blobClient = new BlobClient(apiClient);
    }
    async put(input) {
        return this.blobClient.put({
            ...this.scope,
            ...input,
        });
    }
    async get(input) {
        return this.blobClient.get({
            ...this.scope,
            ...input,
        });
    }
    async delete(input) {
        return this.blobClient.delete({
            ...this.scope,
            ...input,
        });
    }
}
exports.ScopedBlobClient = ScopedBlobClient;
